File: 971-shortest-bridge/shortest-bridge.java
class Solution {
    // Direction array to represent the 4 directions
    private final int[][] directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    int n, m;
    private void dfs(int row, int col, int[][] grid, Queue<int[]> q)
    {
        if(row < 0 || col < 0 || row >= n || col >= m || grid[row][col] != 1)
        return;

        // Mark the first island cell as -1
        grid[row][col] = -1;

        // push in the queue
        q.add(new int[]{row, col, 0});


        for(int[] dir : directions)
        {
            int r = row + dir[0];
            int c = col + dir[1];
            dfs(r, c, grid, q);
        }
    }

    int bfs(int[][] grid, Queue<int[]> q)
    {
        while(!q.isEmpty())
        {
            int[] arr = q.poll();
            int row = arr[0];
            int col = arr[1];
            int dist = arr[2];

            for(int[] dir : directions)
            {
                int r = row + dir[0];
                int c = col + dir[1];
                if(r >= 0 && r < n && c >= 0 && c < m && grid[r][c] >= 0)
                {
                    // Found the first island
                    if(grid[r][c] == 1)
                    return dist;

                    // mark the cell as vis
                    grid[r][c] = -1;

                    // push in the queue
                    q.add(new int[]{r, c, dist + 1});
                }
            }
        }
        // whatever the fuck you want
        return -1;
    }

    public int shortestBridge(int[][] grid) 
    {
        n = grid.length;
        m = grid[0].length;

        Queue<int[]> q = new LinkedList<>();

        // call the dfs from any one island cell
        boolean found = false;
        for(int i = 0; i < n; i++)
        {
            for(int j = 0; j < m; j++)
            if(grid[i][j] == 1)
            {
                dfs(i, j, grid, q);
                found = true;
                break;
            }
            if(found)
            break;
        }

        // Perform BFS to find the shortest bridge
        return bfs(grid, q);
    }
}

File: 971-shortest-bridge/shortest-bridge.cpp
class Solution {
private:
    int n, m;
    vector<pair<int,int>> directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
public:
    void dfs(int row, int col, vector<vector<int>>& grid, vector<vector<bool>>& vis, queue<pair<int,int>>& queue){
        if(row < 0 || col < 0 || row >= n || col >= m || grid[row][col] != 1 || vis[row][col]){
            return;
        }

        vis[row][col] = true;
        queue.push({row, col});

        for(auto dir : directions){
            int newRow = row + dir.first;
            int newCol = col + dir.second;
            dfs(newRow, newCol, grid, vis, queue);
        }
    }

    int shortestBridge(vector<vector<int>>& grid) {

        n = grid.size();
        m = grid[0].size();

        bool canExit = false;
        vector<vector<bool>> vis(n, vector<bool>(m, false));
        queue<pair<int,int>> queue;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(grid[i][j] == 1){
                    canExit = true;
                    dfs(i, j, grid, vis, queue);
                    break;
                }
            }
            if(canExit){
                break;
            }
        }

        int steps = 0;
        while(!queue.empty()){
            int level = queue.size();
            for(int i = 0; i < level; i++){
                auto curr = queue.front();
                queue.pop();
                int row = curr.first;
                int col = curr.second;

                for(auto dir : directions){
                    int newRow = row + dir.first;
                    int newCol = col + dir.second;
                    if(newRow < 0 || newCol < 0 || newRow >= n || newCol >= m || vis[newRow][newCol]){
                        continue;
                    }

                    if(grid[newRow][newCol] == 1){
                        return steps;
                    }

                    queue.push({newRow, newCol});
                    vis[newRow][newCol] = true;
                }
            }
            steps++;
        }
        return -1;
    }
};